### PR TITLE
Net hostlist upgrade

### PR DESCRIPTION
**Pull request by lunar-mining**
This upgrade replaces the net hostlist with a new hostlist consisting of 3 parts:

- `anchorlist`
- `whitelist`
- `greylist`

Hostlists are a vector of tuples that contains a `Url` and a `last_seen` field, which is a timestamp of the last time the host has been interacted with. The lists are ordered according to `last_seen` with the most recent host on the top of the list. The `whitelist` has a max size of 5000 and the `greylist` has a max size of 2000. This is defined in `net/hosts/store.rs`. 

Nodes send `GetAddr` messages to request host information from other peers. On receiving `GetAddr`, nodes reply with an `Addr` message containing their `whitelist`. On receiving `Addr`, nodes insert the information into their `greylist`. 

If nodes have `advertise` set to true, they will also broadcast their own addresses.  Before doing this, they ping themselves using the `ping_node()` method defined in `net/store/refinery.rs` to ensure their external address is reachable before sending it. This behavior is defined in `ProtocolAddr` and `ProtocolSeed`.

Periodically, the  refinery process defined in `net/hosts/refinery.rs` randomly selects an addresses from the `greylist` and tries to do a version exchange with it using `ping_node()`. If the host is responsive, the refinery updates the `last_seen` field and inserts it into the whitelist. Otherwise, it is removed from the `greylist`.

When a connection has been successfully established to another peer, this connection is added to the `anchorlist`. In `OutboundSession`, nodes select first from the `anchorlist`, then from the `whitelist`, then from the `greylist`. The number of `anchorlist` or `whitelist` connections that nodes prefer when making outbound connections is defined in `Settings`.

Previously the hostlist was stored in memory. In this upgrade, all host information (white, grey, anchor) is stored on disk, saved on `refinery.stop()` and loaded on `refinery.start()`.

Lilith has also been updated to reflect the new changes.

### CODE REVIEW

File: bin/lilith/src/main.rs [hunk 1]
```diff
@@ -337,6 +178,11 @@ fn parse_configured_networks(data: &str) -> Result<HashMap<String, NetInfo>> {
                     continue
                 }
 
+                if !table.contains_key("hostlist") {
+                    warn!(target: "lilith", "Hostlist path is mandatory, skipping network.");
```
Comment (parazyd) [inline bin/lilith/src/main.rs:182]: This should be a fatal error instead. Let's not allow misconfigurations.

File: bin/lilith/src/main.rs [hunk 2]
```diff
@@ -521,16 +333,11 @@ async fn realmain(args: Args, ex: Arc<Executor<'static>>) -> Result<()> {
     signals_handler.wait_termination(signals_task).await?;
     info!(target: "lilith", "Caught termination signal, cleaning up and exiting...");
 
-    // Save in-memory hosts to tsv file
-    save_hosts(&expand_path(&args.hosts_file)?, &lilith.networks).await;
-
```
Comment (parazyd) [inline bin/lilith/src/main.rs:526]: This was removed? Should be done upon exit.

File: src/net/hosts/store.rs [hunk 3]
```diff
@@ -0,0 +1,1485 @@
...
+    /// Get the index and entry for a given addr on the anchorlist.
+    pub async fn get_anchorlist_entry_at_addr(&self, addr: &Url) -> Result<(usize, (Url, u64))> {
+        let anchorlist = self.anchorlist.read().await;
+        for (i, (url, time)) in anchorlist.iter().enumerate() {
+            if url == addr {
+                return Ok((i, (url.clone(), time.clone())))
+            }
+        }
+        return Err(Error::HostDoesNotExist)
+    }
```
Comment (parazyd) [inline src/net/hosts/store.rs:792]: All of these can probably be reduced to oneliners

File: src/net/hosts/store.rs [hunk 4]
```diff
@@ -0,0 +1,1485 @@
...
+    //        assert!(hosts.is_empty_whitelist().await);
+    //        assert!(!hosts.is_empty_greylist().await);
+
+    //        let host =
+    //            hosts.anchorlist_fetch_address(p2p.clone(), &transports).await.unwrap();
+
+    //        assert!(grey_urls.iter().any(|u| u.clone() == host.0));
+    //    });
+    //}
+}
```
Comment (parazyd) [inline src/net/hosts/store.rs:1485]: Use `/* */` style comments for big code blocks

File: bin/lilith/src/main.rs (hunk 2 above)
Comment (lunar-mining) [reply bin/lilith/src/main.rs:526]: It now gets done in the refinery, as with all other nodes https://github.com/darkrenaissance/darkfi/blob/2dbaf413a0ed0059e2d8433f231f175207fa9513/src/net/hosts/refinery.rs#L70

Comment (parazyd) [reply bin/lilith/src/main.rs:526]: Ah great

File: src/net/hosts/store.rs (hunk 4 above)
Comment (bobsummerwill) [reply src/net/hosts/store.rs:1485]: Or better yet, delete it all.

That's what version control is for, if you want to readd it later :-)

The commented out stuff is just more cognitive load for everyone.

Comment (lunar-mining) [reply src/net/hosts/store.rs]: I simply forgot to reimplement the tests. Fixed on 639f1f72bf2908837ac1de0cd6baeef907a5ac96

**Comment by parazyd** [comment]
Okay, please fix the tests and clippy lints and I think we can merge it.

**Pull request merged**